server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/SortedDocsProducer.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.composite;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSet;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.DocIdSetBuilder;
+import org.elasticsearch.common.inject.internal.Nullable;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+
+import java.io.IOException;
+
+/**
+ * A producer that visits composite buckets in the order of the value indexed in the leading source of the composite
+ * definition. It can be used to control which documents should be collected to produce the top composite buckets
+ * without visiting all documents in an index.
+ */
+abstract class SortedDocsProducer {
+    protected final String field;
+
+    SortedDocsProducer(String field) {
+        this.field = field;
+    }
+
+    /**
+     * Visits all non-deleted documents in <code>iterator</code> and fills the provided <code>queue</code>
+     * with the top composite buckets extracted from the collection.
+     * Documents that contain a top composite bucket are added in the provided <code>builder</code> if it is not null.
+     *
+     * Returns true if the queue is full and the current <code>leadSourceBucket</code> did not produce any competitive
+     * composite buckets.
+     */
+    protected boolean processBucket(CompositeValuesCollectorQueue queue, LeafReaderContext context, DocIdSetIterator iterator,
+                                    Comparable<?> leadSourceBucket, @Nullable DocIdSetBuilder builder) throws IOException {
+        final int[] topCompositeCollected = new int[1];
+        final boolean[] hasCollected = new boolean[1];
+        int cost = (int) iterator.cost();
+        final DocIdSetBuilder.BulkAdder adder = builder != null ? builder.grow(cost) : null;
```

Comment:
This optimization is unsafe since cost may be inaccurate. I would be ok if it was only called with postings, but it looks like this method is sometimes called with the result of DocIdSetBuilder.finish which gives approximate costs in the dense case.